### PR TITLE
HEIF plugin

### PR DESCRIPTION
**Pull request by david-poirier-csn**
Hey, this is a *DRAFT* version of a HEIF plugin using `pyheif`. Just running it up the flagpole ;)
- Is this wanted?
- Am I on the right track with the plugin design?

If so I'll proceed and add exif, color profile, doco and tests.

#2806

### CODE REVIEW

**Comment by radarhere** [comment]
I think that given the licensing issues, this is the only way forward, and presuming it works, then thanks very much for thinking of this. There is certainly a large number of people requesting this feature over in #2806

@wiredfool, did you have any thoughts on this approach?

**Comment by david-poirier-csn** [comment]
Re: the licensing issues, `libheif` now supports AVIF (using libaom) in addition to HEIC and thus so does `pyheif` (if built from source, the binary wheels don't *yet*). In a nutshell HEIF is the container format, HEIC is a HEIF file containing at least one HEVC (h.265) keyframe, and AVIF is a HEIF file containing at least one AV1 keyframe. Chrome and Firefox are both on the cusp of supporting AVIF. So this is a "twofer" with potential to be even more than that if other image formats adopt the HEIF container format.

On another note I should point out that `pyheif` is Python 3.6+ only, and Linux + MacOS only. I don't know if `libheif` can be built on Windows without significant effort.

**Comment by hugovk** [comment]
> On another note I should point out that `pyheif` is Python 3.6+ only, and Linux + MacOS only. I don't know if `libheif` can be built on Windows without significant effort.

Python 3.5 will be EOL and dropped in the next release: https://github.com/python-pillow/Pillow/pull/4746.

**Comment by nulano** [comment]
> I don't know if libheif can be built on Windows without significant effort.

A brief search turned up with an [ImageMagick libheif repository](https://github.com/ImageMagick/libheif) which claims to be intended for building libheif on Windows, so it should be possible:

> ## ImageMagick / libheif
> ### About
> libheif for the Windows build of ImageMagick

If this gets merged, I am willing to *attempt* to contribute scripts for Windows builds to pyheif (e.g. using GitHub Actions).

**Comment by jolyonralph** [comment]
Just a quick comment to the first question you asked - yes it IS wanted! 

Thank you!

File: src/PIL/HeifImagePlugin.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    return check != pyheif.heif_filetype_no
+
+
+class HeifImageFile(ImageFile.ImageFile):
+    format = "HEIF"
+    format_description = "HEIF Image File Format"
+
+    def _open(self):
+        b = self.fp.read()
+        if not _accept(b):
```
Comment (wiredfool) [inline src/PIL/HeifImagePlugin.py:21]: In general we should not be reading the whole file in _open. It does appear that pyheif does that, which is somewhat disappointing. I'm not sure if it's possible to get around that, except for the initial magic check.

File: src/PIL/HeifImagePlugin.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            raise ValueError("Not a HEIF image file")
+
+        heif_file = pyheif.read(b)
+        self._size = heif_file.size
+        self.mode = heif_file.mode
+        self.fp = io.BytesIO(heif_file.data)
```
Comment (wiredfool) [inline src/PIL/HeifImagePlugin.py:27]: The contract on _open is that we don't do expensive operations, so decoding the entire images and storing a copy violates that. 

Unfortunately, it doesn't appear that pyheif makes it easy to get the metadata that we require for the _open command without decoding the entire image.

**Review by wiredfool — COMMENTED** [review]
Sorry this is a stale review. 2020. 

I think this is a reasonable approach from a licensing point of view, and it should be shippable that way. I'm not sure that we'd ever get libheif in the binary builds, but this would be the sort of thing that would work at arms length if someone were to install pyheif separately. 

I'm a concerned with the amount of work that's done in open. I think it might be worth seeing if the pyheif project would be amenable to having a basic call into the library for size/mode/etc without the load of a full decode.

File: src/PIL/HeifImagePlugin.py (hunk 2 above)
Comment (david-poirier-csn) [reply src/PIL/HeifImagePlugin.py:27]: Thanks, I thought I did have a metadata-only call but maybe I'm not using it. I'll check it out.

**Comment by myers** [comment]
See https://pypi.org/project/pyheif-pillow-opener/

**Comment by david-poirier-csn** [comment]
Haha yes, I think it will be easier to register like this from `pyheif`, rather than try to integrate with `Pillow` due to H.265 patent issues etc

I will still try to implement @wiredfool's advice, much appreciated

**Comment by homm** [comment]
To whom it may concern. I've made [heif-image-plugin](https://pypi.org/project/heif-image-plugin/) where fixed many issues including described here. For example:

* `im.info['icc_profile']` is now accessible 
* works without decoding whole image on open
* Opens AVIF files as well

**Pull request closed without merging**